Fourth Batch for PubSub Support

andreasebner opened the pull request:
Add PubSub WriterGroups (WG), DataSetWriters (DSW) and DataSetField  (DSF) handling. The WriterGroups are contained in the PubSubConnections. The DataSetField is contained in the PublishedDataSet and selects the published information. The DataSetWriters are the glue between the PublishedDataSet and the WriterGroups. The are used to configure the message creation.

This introduces the following API changes:
NEW
UA_[WG/DSW/DSF]_getConfig: get the current config
UA_[WG/DSW/DSF]_deleteMembers:  delete the config and all nested members
UA_PubSubConnection_addWriterGroup: add WG to PubSubConnection
UA_PubSubConnection_removeGroup: remove WG from PubSubConnection + nested members
UA_PublishedDataSet_addDataSetField: add DSF to PDS
UA_PublishedDataSet_removeField: remove DSF from PDS + nested members
UA_WriterGroup_addDataSetWriter: add DSW to WG
UA_WriterGroup_removeDataSetWriter: remove DSW from WG + nested members


More details about PDS handling can be found in tutorials_pubsub_publisher

jpfr commented:
Hey @andreasebner,

the code is very clean and organized. I mostly focused on the public API for the review.

I have only cosmetic changes.
I made some changes that live in this branch. Please squash/fixup if you agree.
https://github.com/open62541/open62541/tree/pubsub_pr4

- Moved the type definition update to a separate commit.
- Rename methods to UA_Server_... in the public API
- Use const nodeids where possible
- Move doc-comments from .c to .h

One thing I could not answer right away: Here you introduce `DataSetField`. Are these the same for Published and Subscriber? Otherwise, should we call them `PublishedDataSetField` to avoid future name clashes?

andreasebner commented:
Hey @jpfr,

thanks for you review and code adjustments. I have now rebased your changes and fixed some parameter renaming issues. 

I took a look into the standard to check your naming remark. The SubscribedDataSet contains a FieldTargetDataType. In order to the other naming conventions we should use a function name like 'addFieldTarget' for the subscriber. 
But if you want to avoid any problems, i can refactor the code and change DataSetField to PublishedDataSetField ;)

jpfr commented:
👍
